File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/New/Future/SubSystems/LinearSlides.java
package org.firstinspires.ftc.teamcode.New.Future.SubSystems;

import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.HardwareMap;

import org.apache.tools.ant.taskdefs.Tar;
import org.firstinspires.ftc.teamcode.New.PIDFcontroller;
import org.firstinspires.ftc.teamcode.New.PIDParams;

public class LinearSlides {

    public LeftSlide leftSlide;
    public RightSlide rightSlide;
    public int resetValue = 0;

    public void init(){
        leftSlide.leftSlide.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        rightSlide.rightSlide.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
    }

    public LinearSlides(HardwareMap hardwareMap){
        leftSlide = new LeftSlide(hardwareMap);
        rightSlide = new RightSlide(hardwareMap);
    }
    public enum State{
        HANG(5000),
        BASKET(430),
        CLIP(340),
        INTAKE(550),
        STATIONARY(220),
        IDLE(0),
        Manual(0);
        State(int newEncoder){
            Encoder = newEncoder;
        }
        final int Encoder;
    }

    public State state = State.IDLE;
    int target = 0;

    public boolean hasReached(){
        int leftError = target - leftSlide.leftEncoder;
        int rightError = target - rightSlide.rightEncoder;
        int averageError = (rightError + leftError)/2;
        double tolerance = 0.01;

        return averageError > -tolerance && averageError < tolerance;
    }
    public static int manualTarget = 0;

    public void update(){
        target = state.Encoder;
        if(state == State.Manual){
            target = manualTarget;
        }
        leftSlide.update(target, state);
        rightSlide.update(target, state);
    }

    public static class LeftSlide {

        DcMotor leftSlide;
        public int leftEncoder;

        //tune PID coefficients
        PIDParams pidParams = new PIDParams(0.0,0.0,0.0,0.0);
        PIDFcontroller pidFcontroller = new PIDFcontroller(pidParams);

        public LeftSlide(HardwareMap hardwareMap){
            leftSlide = hardwareMap.get(DcMotor.class, "leftSlide");
        }


        public void update(int target, State state) {
            leftEncoder = leftSlide.getCurrentPosition();

            double leftPower = pidFcontroller.calculate(target - leftEncoder);

            //idle mode for testing or emergency stop
            if(state == State.IDLE){
               leftPower = 0;
           }

            leftSlide.setPower(leftPower);
        }
    }

    public static class RightSlide {
        DcMotor rightSlide;
        public int rightEncoder;

        //tune PID coefficients
        PIDParams pidParams = new PIDParams(0.0,0.0,0.0,0.0);
        PIDFcontroller pidFcontroller = new PIDFcontroller(pidParams);

        public RightSlide(HardwareMap hardwareMap){
            rightSlide = hardwareMap.get(DcMotor.class, "rightSlide");
        }

        public void update(int target, State state) {
            rightEncoder = rightSlide.getCurrentPosition();

            double rightPower = pidFcontroller.calculate(target - rightEncoder);

            //idle mode for testing or emergency stop
            if(state == State.IDLE){
                rightPower = 0;
            }

            rightSlide.setPower(rightPower);
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/New/Heisenberg/Subsystems/LinearSlides.kt
package org.firstinspires.ftc.teamcode.New.Heisenberg.Subsystems

import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.robotcore.external.navigation.CurrentUnit
import org.firstinspires.ftc.teamcode.New.Utilities.Controller
import org.firstinspires.ftc.teamcode.New.Utilities.PIDParams

class LinearSlides(hwMap:HardwareMap) {

    val motor: DcMotorEx = hwMap.get(DcMotorEx::class.java, "slide")
    init {
        motor.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        motor.mode  = DcMotor.RunMode.RUN_WITHOUT_ENCODER
        motor.direction = DcMotorSimple.Direction.REVERSE
    }

    //todo: Create an opmode that uses this subsytem - and sets the pid controller's paramters to ftc Dashboard numbers

    private val pidController = Controller(PIDParams(.0085,0.0,0.0007,0.0))


    //NOTE: Max encoder ticks is 311, otherwise we break hardware

//    private val circumferenceOfSpool: Double = 24 * Math.PI
//    var conv: SlidesEncoderConv = SlidesEncoderConv(circumferenceOfSpool)

    var state = SlidesState.IDLE
    private var offset = 0
    private var currentPos =0

    fun update(){
        //in to tick
//        if(motor.getCurrent(CurrentUnit.AMPS) >= 9.0 && currentPos <= 300) offset+= motor.currentPosition
        currentPos = motor.currentPosition - offset
        motor.power = pidController.calculate(state.distance -currentPos)
    }

//    private fun InToTick(input: Double): Double {
//        return 0.0
//    }


    enum class SlidesState(val distance: Double) {
        WALL(0.0),
        INTAKE(540.0),
        BASKET(880.0),
        Specimen(700.0),
        SpecimenHang(650.0),
        IDLE(0.0),
        HANG(700.0);
    }

}